### PR TITLE
Fix up Time/Duration handling.

### PR DESCRIPTION
**Pull request by jlebar**
<git-pr-chain>

#### Commits in this PR
1. Fix up Time/Duration handling.
    
    - Make Time an int64 and handle clock overflow.
    - Remove weakly-typed Hal::millis() and Hal::delay(int32) functions.
    - Fix up code so that it all uses strongly-typed functions.
    
    The main motivation is handling timer overflow properly.  As we saw in
    https://github.com/RespiraWorks/VentilatorSoftware/issues/224, overflow
    can be very serious!

#### [PR chain](https://github.com/jlebar/git-pr-chain)
1. 👉 #238 Fix up Time/Duration handling. 👈 **YOU ARE HERE**
1. #205 Clarify comment on sensor scaling factor.

</git-pr-chain>

### CODE REVIEW

File: controller/lib/hal/hal.h [hunk 1]
```diff
@@ -266,7 +257,7 @@ class HalApi {
   void setDigitalPinMode(BinaryPin pin, PinMode mode);
 
 #ifdef TEST_MODE
-  uint32_t millis_ = 0;
+  int64_t millis_ = 0;
```
Comment (jkff) [inline controller/lib/hal/hal.h:260]: I guess this could be a Time too

File: controller/lib/hal/hal.h [hunk 2]
```diff
@@ -283,6 +274,16 @@ class HalApi {
...
 #endif
+
+#if defined(ARDUINO_AVR_UNO) || defined(ARDUINO_NUCLEO_L452RE)
+  // Value of ::millis() at last call to watchdog_handler().  Used to detect
+  // clock overflows.
+  uint32_t last_millis_ = 0;
+
+  // Number of times millis() has gone backwards.   Arduino millis is a uint32,
+  // so this happens once every 1.6 months of uptime.
+  uint32_t clock_overflows_ = 0;
```
Comment (jkff) [inline controller/lib/hal/hal.h:285]: num_clock_overflows_?

File: controller/lib/hal/hal.h [hunk 3]
```diff
@@ -402,15 +406,21 @@ inline void HalApi::watchdog_init() {
 #endif
 }
 inline void HalApi::watchdog_handler() {
+  uint32_t millis = ::millis();
```
Comment (jkff) [inline controller/lib/hal/hal.h:409]: Explain why we do this in the watchdog handler?

File: controller/lib/hal/hal.h (hunk 2 above)
Comment (jlebar) [reply controller/lib/hal/hal.h:285]: Done

File: controller/lib/hal/hal.h (hunk 3 above)
Comment (jlebar) [reply controller/lib/hal/hal.h:409]: Done.

File: controller/lib/hal/hal.h (hunk 1 above)
Comment (jlebar) [reply controller/lib/hal/hal.h:260]: Oh, sure.

**Comment by jlebar** [comment]
Thank you, Eugene.

**Pull request merged**